File: 프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.py
def solution(arr):
    counts = [0, 0]
    
    def compress(r, c, size):
        first_elem = arr[r][c]
        same = True
        
        for i in range(r, r+size):
            for j in range(c, c+size):
                if arr[i][j] != first_elem:
                    same = False
                    break
                if not same:
                    break
        if same:
            if first_elem == 0:
                counts[0] += 1
            else:
                counts[1] += 1
            return
        
        half = size // 2
        compress(r, c, half)
        compress(r+half, c, half)
        compress(r, c+half, half)
        compress(r+half, c+half, half)
        
    compress(0, 0, len(arr))
    return counts
                    

File: 프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.js
function solution(arr) {
    let zeros = 0;
    let ones = 0;
    
    function compress(r, c, size){
        // 배열 안의 모든 숫자가 같을 때
        let same = true;
        let firstElem = arr[r][c];
        check: for(let i=r;i<r+size;i++){
            for(let j=c;j<c+size;j++){
                if(firstElem !== arr[i][j]){
                    same = false;
                    break check;
                }
            }
        }
        if(same){
            firstElem === 0 ? zeros++ : ones++;
            return;
        }
        // 배열 안의 모든 숫자가 같지 않으면 넷으로 나누어 반복
        const half = size / 2;
        compress(r, c, half);
        compress(r+half, c, half);
        compress(r, c+half, half);
        compress(r+half, c+half, half);
    }
    compress(0, 0, arr.length);
    return [zeros, ones];
}